Improve memory performance on publishing

fabricio-aguiar opened the pull request:
https://pulp.plan.io/issues/5689
closes #5689

fabricio-aguiar commented on pulp_rpm/app/models/comps.py:
```diff
@@ -107,10 +107,14 @@ def pkglist_to_lst(cls, value):
         """
         package_list = []
         for i in value:
-            package_list.append({'name': i.name,
```
I don't know why, but sometimes it duplicates some packages

fabricio-aguiar commented on pulp_rpm/app/tasks/publishing.py:
```diff
@@ -58,8 +57,6 @@ def __init__(self, publication):
         """
         self.publication = publication
-        self.packages = []
-        self.published_artifacts = []
```
it was holding objects in memory

fabricio-aguiar commented on pulp_rpm/app/tasks/publishing.py:
```diff
@@ -95,21 +92,46 @@ def prepare_metadata_files(self, contents, folder=None):
-            packages (pulp_rpm.models.Package): A list of packages.
+        """
+        published_artifacts = []
+        for content_artifact in ContentArtifact.objects.filter(content__in=contents).iterator():
```
instead of going to each kind of content and prefetch its `content_artifact`, it was better to use `ContentArtifact` for all contents, using `iterator`, so it would not cache all the data in memory

fabricio-aguiar commented on pulp_rpm/app/tasks/publishing.py:
```diff
@@ -132,48 +155,20 @@ def populate(self):
-            sub_repo_packages = self.get_packages(content)
-            all_packages = all_packages | sub_repo_packages
-            setattr(self, f"{name}_packages", sub_repo_packages)
+            setattr(self, f"{name}_contents", content)
```
exposing `contents` instead of `packages` this change is due to kickstarts, I will explain better on the next comments

fabricio-aguiar commented on pulp_rpm/app/tasks/publishing.py:
```diff
@@ -246,14 +241,17 @@ def create_repomd_xml(packages, publication, extra_repomdrecords, sub_folder=Non
     oth_db = cr.OtherSqlite(oth_db_path)
     upd_xml = cr.UpdateInfoXmlFile(upd_xml_path)
 
-    pri_xml.set_num_of_pkgs(len(packages))
```
it was bringing the entire query to memory, for 27000 packages, it was huge!

fabricio-aguiar commented on pulp_rpm/app/tasks/publishing.py:
```diff
@@ -246,14 +241,17 @@ def create_repomd_xml(packages, publication, extra_repomdrecords, sub_folder=Non
 
     # Process all packages
-    for package in packages:
+    for package in packages.iterator():
```
By default, [iterator](https://docs.djangoproject.com/en/2.2/ref/models/querysets/#iterator) only bring 2000 rows into memory, as we only would use packages on this loop, it was better to use iterator than caching the entire query.

fabricio-aguiar commented on pulp_rpm/app/tasks/publishing.py:
```diff
@@ -262,40 +260,33 @@ def create_repomd_xml(packages, publication, extra_repomdrecords, sub_folder=Non
     # Process update records
-    for update_record in UpdateRecord.objects.filter(
-            pk__in=publication.repository_version.content):
+    for update_record in UpdateRecord.objects.filter(pk__in=contents).iterator():
```
that's the reason why I exposed `contents` instead of `packages`. Before, this and the other queries below were only pointing to main repo contents. It was a bug for kickstarts, because sub repo contents were not exposed, so this query would never find the desired sub repo content

fabricio-aguiar commented on pulp_rpm/app/tasks/publishing.py:
```diff
@@ -95,21 +92,46 @@ def prepare_metadata_files(self, contents, folder=None):
         return repomdrecords
 
-    def get_packages(self, contents):
+    def publish_artifacts(self, contents):
```
actually I emphasize `contents` to bring the attention it is not just one thing, it can be huge

goosemania commented on pulp_rpm/app/tasks/publishing.py:
```diff
@@ -95,21 +92,46 @@ def prepare_metadata_files(self, contents, folder=None):
 
+        PublishedArtifact.objects.bulk_create(published_artifacts, batch_size=2000)
+
+    def handle_sub_repo(self, lst, identifier):
```
And what does it do? It only publishes artifacts, not metadata, correct?
how about `publish_sub_repo_artifacts`?

goosemania replied:
I think it would be clearer for future code readers if the names are more descriptive.
How about `lst`->`sub_repos` and `identifier` -> `sub_repo_type`?

fabricio-aguiar replied:
it not only publish artifacts, it also makes sub repo content available to be used on `create_repomd_xml`

fabricio-aguiar replied:
identifier would be `addon_id` or `variant_id`, I don't know if `sub_repo_type` suits well

goosemania commented on pulp_rpm/app/tasks/publishing.py:
```diff
@@ -95,21 +92,46 @@ def prepare_metadata_files(self, contents, folder=None):
-        return packages
+        """
+        sub_repos_pks = []
+        for item in lst:
```
same here, can we make it more descritpive, e.g. `item` -> `sub_repo`?

fabricio-aguiar replied:
item would be addon/variant, I don't know which generic name could cover it

goosemania commented on pulp_rpm/app/tasks/publishing.py:
```diff
@@ -95,21 +92,46 @@ def prepare_metadata_files(self, contents, folder=None):
+            if repository_version and repository.sub_repo:
+                item_id = getattr(item, identifier)
+                self.sub_repos.append((item_id, repository_version.content))
+                if repo_pk not in sub_repos_pks:
```
do you expect duplicated sub_repos?

fabricio-aguiar replied:
as worked with few multi-addons/variants, I'm just being cautious. Answering in another way, I'm not sure what to expect

daviddavis commented:
Everything here looks good to me with the exception of the `handle_sub_repo()` function. It seems to add a lot of complexity over what we currently have and I'm not quite clear on how it improves memory usage?

fabricio-aguiar commented:
@daviddavis and @goosemania `handle_sub_repo` is this:
https://github.com/pulp/pulp_rpm/blob/master/pulp_rpm/app/tasks/publishing.py#L141-L148
just a DRY nitpick

fabricio-aguiar commented:
> @daviddavis and @goosemania `handle_sub_repo` is this:
> https://github.com/pulp/pulp_rpm/blob/master/pulp_rpm/app/tasks/publishing.py#L141-L148
> just a DRY nitpick

actually it was not only a DRY nitpick, 
I changed how to determine if it is a sub-repo from:
```python
if repository_version and repository_version.content != main_content:
```
to
```python
if repository_version and repository.sub_repo:
```
comparing content was more expensive than using sub_repo attribute

and also introduced some checking if the content was not published, if not, then publish it
